### PR TITLE
Implement Storage bucket IAM methods

### PR DESCRIPTION
**Pull request by jskeet**
Reasonable to review either by commit, or all in one go. If you review by commit, be aware that the final commit fixes bugs in the first two commits :)

### CODE REVIEW

**Comment by jskeet** [comment]
Note: this feature is currently behind a whitelist, so we may well want to hold on until it's GA before merging.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/GetBucketIamPolicyOptions.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (evildour) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/GetBucketIamPolicyOptions.cs]: 2017

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/SetBucketIamPolicyOptions.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (evildour) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/SetBucketIamPolicyOptions.cs]: 2017

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/StorageClient.SetBucketIamPolicy.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Storage.V1
+{
+    public abstract partial class StorageClient
+    {
+        /// <summary>
+        /// Fetches the IAM policy information about a bucket synchronously.
```
Comment (evildour) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/StorageClient.SetBucketIamPolicy.cs]: This summary should be updated for the set operation.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/StorageClient.SetBucketIamPolicy.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        /// <param name="options">Additional options for the operation. May be null, in which case appropriate
+        /// defaults will be used.</param>
+        /// <returns>The <see cref="Policy"/> representation of the resulting storage bucket IAM policy.</returns>
+        public virtual Policy SetBucketIamPolicy(string bucket, Policy policy, SetBucketIamPolicyOptions options = null)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Fetches the IAM policy information about a bucket asynchronously.
```
Comment (evildour) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/StorageClient.SetBucketIamPolicy.cs]: Same here.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/StorageClient.TestBucketIamPermissions.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        /// <param name="options">Additional options for the operation. May be null, in which case appropriate
+        /// defaults will be used.</param>
+        /// <returns>The permissions held by the caller.</returns>
+        public virtual IList<string> TestBucketIamPermissions(string bucket, IEnumerable<string> permissions, TestBucketIamPermissionsOptions options = null)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Fetches the IAM policy information about a bucket asynchronously.
```
Comment (evildour) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/StorageClient.TestBucketIamPermissions.cs]: This summary should be updated.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/TestBucketIamPermissionsOptions.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (evildour) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/TestBucketIamPermissionsOptions.cs]: 2017

**Comment by jskeet** [comment]
Thanks for all these Mike - will fix them all.

**Comment by jskeet** [comment]
Commits are now sensible again. We may well remove the final commit, depending on whether we settle on Update or Set as the verb...

**Comment by jskeet** [comment]
(That commit removed now, as Set has been chosen as the verb.)

**Comment by evildour** [comment]
LGTM

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Snippets/StorageClientSnippets.cs [hunk 7]
```diff
@@ -594,5 +596,101 @@ public void DeleteBucket()
...
+                binding = new Policy.BindingsData { Role = role, Members = new List<string>() };
+                policy.Bindings.Add(binding);
+            }
+            binding.Members.Add("allUsers");
+
+            // Update the IAM policy on the bucket.
+            client.SetBucketIamPolicy(bucketName, policy);
+
+            // Download the object again: this time the response should be OK
+            HttpResponseMessage response2 = await httpClient.GetAsync(obj.MediaLink);
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Snippets/StorageClientSnippets.cs]: Why is the URL being retrieved different from the first attempt, bucket.SelfLink ? Shouldn't it be the same link, to demonstrate that only the bucket IAM permissions need to change?

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Snippets/StorageClientSnippets.cs [hunk 8]
```diff
@@ -594,5 +596,101 @@ public void DeleteBucket()
...
+            Console.WriteLine($"Response code after setting policy: {response2.StatusCode}");
+
+            // End snippet
+
+            Assert.Equal(HttpStatusCode.Unauthorized, response1.StatusCode);
+            Assert.Equal(HttpStatusCode.OK, response2.StatusCode);
+        }
+
+        // See-also: SetBucketIamPolicy(string, *, *)
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Snippets/StorageClientSnippets.cs]: nit: Spacing inconsistency with other See-also comments.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Snippets/StorageClientSnippets.cs (hunk 7 above)
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Snippets/StorageClientSnippets.cs]: I should use MediaLink for both - will fix.

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Snippets/StorageClientSnippets.cs (hunk 8 above)
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Snippets/StorageClientSnippets.cs]: Indeed, I've only recently started permitting spaces in the tooling. I think it's clearer to read, but I haven't gone back and redone things. I may do that in a separate PR...

File: apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Snippets/StorageClientSnippets.cs (hunk 7 above)
Comment (jskeet) [inline apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Snippets/StorageClientSnippets.cs]: Oops - I hadn't spotted that we weren't retrieving the *object* at all in the first case. All done now though, and it still passes :) Will merge on green.
